File: secrethub/lib/secrethub/feature_hub_provider.ex
defmodule Secrethub.FeatureHubProvider do
  use FeatureProvider.Provider
  require Logger

  alias InternalApi.Feature.{
    OrganizationFeature,
    Availability
  }

  @impl FeatureProvider.Provider
  def provide_features(org_id, _opts) do
    alias InternalApi.Feature.ListOrganizationFeaturesRequest, as: Request
    alias InternalApi.Feature.FeatureService.Stub

    Watchman.benchmark("feature_hub.list_organization_features.duration", fn ->
      req = Request.new(org_id: org_id)

      {:ok, ep} = Application.fetch_env(:secrethub, :feature_api_endpoint)

      {:ok, ch} = GRPC.Stub.connect(ep)

      Logger.debug("Listing features for org #{org_id} from GRPC endpoint #{inspect(ep)}")

      Stub.list_organization_features(ch, req, timeout: 3_000)
      |> tap(fn
        {:ok, _res} ->
          Watchman.increment("feature_hub.list_organization_features.success")

        {:error, _} ->
          Watchman.increment("feature_hub.list_organization_features.failure")
      end)
      |> case do
        {:ok, res} ->
          features =
            res.organization_features
            |> Enum.map(&feature_from_grpc/1)

          {:ok, features}

        {:error, e} ->
          Logger.error("Error fetching organization features: #{inspect(e)}")

          {:ok, []}
      end
    end)
  end

  @impl FeatureProvider.Provider
  def provide_machines(_org_id, _) do
    {:ok, []}
  end

  defp feature_from_grpc(%OrganizationFeature{feature: feature, availability: availability}) do
    %FeatureProvider.Feature{
      name: feature.name,
      type: feature.type,
      description: feature.description,
      quantity: quantity_from_availability(availability),
      state: state_from_availability(availability)
    }
  end

  defp quantity_from_availability(%Availability{quantity: quantity}) do
    quantity
  end

  defp state_from_availability(%Availability{state: state}) do
    state
    |> case do
      :ENABLED -> :enabled
      :HIDDEN -> :disabled
      :ZERO_STATE -> :zero_state
    end
  end
end

File: mcp_server/pkg/feature/feature_hub_provider.go
// Package feature holds the feature provider interface and its implementations.
package feature

import (
	"context"
	"errors"
	"net"
	"time"

	"google.golang.org/grpc"

	"github.com/renderedtext/go-watchman"
	pb "github.com/semaphoreio/semaphore/mcp_server/pkg/internal_api/feature"
)

type FeatureHubProvider struct {
	grpcEndpoint string
	dialContext  DialContextFunc
}

type DialContextFunc func(ctx context.Context, addr string) (net.Conn, error)

type FeatureHubProviderOption func(*FeatureHubProvider)

// WithDialContext overrides the default transport dialer used by the provider.
// It is mainly intended for tests where an in-memory listener (for example
// google.golang.org/grpc/test/bufconn) replaces a real network connection.
func WithDialContext(dialer DialContextFunc) FeatureHubProviderOption {
	return func(p *FeatureHubProvider) {
		p.dialContext = dialer
	}
}

func NewFeatureHubProvider(grpcEndpoint string, opts ...FeatureHubProviderOption) (*FeatureHubProvider, error) {
	if grpcEndpoint == "" {
		return nil, errors.New("FeatureHub configuration is invalid, missing grpc endpoint is not set")
	}
	provider := &FeatureHubProvider{grpcEndpoint: grpcEndpoint}
	for _, opt := range opts {
		opt(provider)
	}
	return provider, nil
}

func (p *FeatureHubProvider) ListFeatures(orgID string) ([]OrganizationFeature, error) {
	return p.ListFeaturesWithContext(context.Background(), orgID)
}

func (p *FeatureHubProvider) ListFeaturesWithContext(ctx context.Context, orgID string) ([]OrganizationFeature, error) {
	defer watchman.Benchmark(time.Now(), "feature_hub.list_organization_features.duration")
	conn, err := grpc.DialContext(ctx, p.grpcEndpoint, p.dialOptions()...)
	if err != nil {
		return nil, err
	}

	defer conn.Close()

	client := pb.NewFeatureServiceClient(conn)
	req := pb.ListOrganizationFeaturesRequest{OrgId: orgID}

	response, err := client.ListOrganizationFeatures(ctx, &req)
	if err != nil {
		_ = watchman.Increment("feature_hub.list_organization_features.failure")
		return nil, err
	}

	organizationFeatures := make([]OrganizationFeature, 0, len(response.OrganizationFeatures))
	for _, orgFeature := range response.OrganizationFeatures {
		organizationFeatures = append(organizationFeatures, OrganizationFeature{
			Name:     orgFeature.Feature.Type,
			Quantity: orgFeature.Availability.Quantity,
			State:    stateOfFeature(orgFeature),
		})
	}

	_ = watchman.Increment("feature_hub.list_organization_features.success")
	return organizationFeatures, nil
}

func stateOfFeature(orgFeature *pb.OrganizationFeature) State {
	switch orgFeature.Availability.State {
	case pb.Availability_ENABLED:
		return Enabled
	case pb.Availability_HIDDEN:
		return Hidden
	case pb.Availability_ZERO_STATE:
		return ZeroState
	default:
		return Hidden
	}
}

func (p *FeatureHubProvider) dialOptions() []grpc.DialOption {
	opts := []grpc.DialOption{grpc.WithInsecure()}
	if p.dialContext != nil {
		opts = append(opts, grpc.WithContextDialer(p.dialContext))
	}
	return opts
}
